File: 1609-even-odd-tree/1609-even-odd-tree.ts
/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     val: number
 *     left: TreeNode | null
 *     right: TreeNode | null
 *     constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.left = (left===undefined ? null : left)
 *         this.right = (right===undefined ? null : right)
 *     }
 * }
 */

function bfs(root: TreeNode): Array<Array<number>> {
	let q = [root]
	let answer: Array<Array<number>> = []

	while (q.length > 0) {
		const next_q = new Array<TreeNode>()
		answer.push([])

		for (const it of q) {
			answer[answer.length - 1].push(it.val)
			if (!!it.left) {next_q.push(it.left)}
			if (!!it.right) {next_q.push(it.right)}
		}
		q = next_q
	}

	return answer
}

function is_evenodd(levels: number[][]): boolean {
	for (let i = 0; i < levels.length; i++) {
		const op: (l: number, r: number) => boolean =
			(i & 1) ?
				(l, r) => l > r :
				(l, r) => l < r
		const values = levels[i]

		if (!values.every(it => !!(i & 1) !== !!(it & 1))) {
			return false
		}

		for (let j = 0; j < values.length - 1; ++j) {
			if (!op(values[j], values[j + 1])) {
				return false
			}
		}
	}
	return true
}

function isEvenOddTree(root: TreeNode): boolean {
	const levels = bfs(root)
	return is_evenodd(levels)
};

File: 1609-even-odd-tree/1609-even-odd-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def bfs(self, root):
        q = [(0, root)]
        levels = defaultdict(list)
        while q:
            next_q = []
            for lv, it in q:
                levels[lv].append(it.val)
                if it.left: next_q.append((lv + 1, it.left))
                if it.right: next_q.append((lv + 1, it.right))
            q = next_q
        return levels
    
    def evenodd(self, lv, values):
        if lv & 1:  # decrease
            return all(it % 2 == 0 for it in values) and all(
                it > jt for it, jt in zip(values, values[1:])
            )
        else:  # increase
            return all(it & 1 for it in values) and all(
                it < jt for it, jt in zip(values, values[1:])
            )
    def isEvenOddTree(self, root: Optional[TreeNode]) -> bool:
        levels = self.bfs(root)

        return all(
            self.evenodd(lv, values) for lv, values in levels.items()
        )
